src/util/timer.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::mem::replace;
+use std::cmp::{Ord, Ordering};
+use std::time::Duration;
+use std::collections::BinaryHeap;
+
+use util::time::Instant;
+
+pub struct Timer<T> {
+    pending: BinaryHeap<TimeoutTask<T>>,
+}
+
+impl<T> Timer<T> {
+    pub fn new(capacity: usize) -> Self {
+        Timer {
+            pending: BinaryHeap::with_capacity(capacity),
+        }
+    }
+
+    /// Add a periodic task into the `Timer`.
+    pub fn add_task(&mut self, timeout: Duration, task: T) {
+        let task = TimeoutTask {
+            next_tick: Instant::now() + timeout,
+            timeout: timeout,
+            task: task,
+        };
+        self.pending.push(task);
+    }
+
+    /// Remove a task from the timer. Returns the `TimeoutTask` if found.
+    ///
+    /// If there are many tasks according with the condition, only the first
+    /// task will be removed from pending queue.
+    pub fn remove_task<F>(&mut self, f: F) -> Option<TimeoutTask<T>>
+    where
+        F: Fn(&TimeoutTask<T>) -> bool,
+    {
+        let mut res = None;
+        let mut vec = replace(&mut self.pending, BinaryHeap::new()).into_vec();
+        if let Some((idx, _)) = vec.iter().enumerate().find(|&(_, task)| f(task)) {
+            res = Some(vec.swap_remove(idx));
+        }
+        self.pending = BinaryHeap::from(vec);
+        res
+    }
+
+    /// Get the next `timeout` from the timer.
+    pub fn next_timeout(&mut self) -> Option<Instant> {
+        self.pending.peek().map(|task| task.next_tick)
+    }
+
+    /// Pop a `TimeoutTask` from the `Timer`, which should be tick before `instant`.
+    /// If there is no tasks should be ticked any more, None will be returned.
+    ///
+    /// The normal use case is keeping `pop_task_before` until get `None` in order
+    /// to retreive all avaliable events.
+    pub fn pop_task_before(&mut self, instant: Instant) -> Option<TimeoutTask<T>> {
+        match self.pending.pop() {
```

Comment:
I think you can use `self.pending.peek().map_or(false, |t| t.next_tick > instant)`, so task don't have to be pushed back to heap, which is an expensive operation.